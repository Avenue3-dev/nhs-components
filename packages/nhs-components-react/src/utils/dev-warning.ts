const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
