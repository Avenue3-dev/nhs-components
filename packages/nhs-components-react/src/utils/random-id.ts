export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
