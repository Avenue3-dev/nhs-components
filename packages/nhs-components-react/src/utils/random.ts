export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
