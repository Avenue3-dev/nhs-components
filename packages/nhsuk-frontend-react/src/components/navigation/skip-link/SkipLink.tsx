import React, { useEffect } from 'react';
import { ElementProps } from '@/types/shared';
import clsx from 'clsx';
import initSkipLink from 'nhsuk-frontend/packages/components/skip-link/skip-link';
import { Factory, factory } from '@/internal/factory/factory';

export type SkipLinkProps = ElementProps<'a', 'href'>;

type SkipLinkFactory = Factory<{
  props: SkipLinkProps;
  ref: HTMLAnchorElement;
}>;

const SkipLink = factory<SkipLinkFactory>(
  ({ children, className, ...props }, ref) => {
    useEffect(() => {
      setTimeout(() => {
        initSkipLink();
      }, 500);
    });

    return (
      <a
        className={clsx('nhsuk-skip-link', className)}
        href="#maincontent"
        {...props}
        ref={ref}
      >
        {children}
      </a>
    );
  },
);

SkipLink.displayName = 'SkipLink';

export { SkipLink };
