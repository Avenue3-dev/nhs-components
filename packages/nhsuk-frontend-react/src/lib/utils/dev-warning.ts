const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
