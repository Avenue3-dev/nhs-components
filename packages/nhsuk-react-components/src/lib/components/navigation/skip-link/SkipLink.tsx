import clsx from 'clsx'
import { HTMLProps } from 'react'

/**
 * Use a skip link to help keyboard-only users skip to the main content on a page.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/skip-link).
 *
 * @link https://service-manual.nhs.uk/design-system/components/skip-link
 *
 * @example
 * ```tsx
 * <SkipLink href="#maincontent">Skip to main content</SkipLink>
 * ```
 */
const SkipLink: React.FC<HTMLProps<HTMLAnchorElement>> = ({
  children = 'Skip to main content',
  className,
  href = '#maincontent',
  ...rest
}) => {
  return (
    <a className={clsx('nhsuk-skip-link', className)} href={href} {...rest}>
      {children}
    </a>
  )
}

export default SkipLink
