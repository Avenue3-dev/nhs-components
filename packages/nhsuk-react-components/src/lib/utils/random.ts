export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
