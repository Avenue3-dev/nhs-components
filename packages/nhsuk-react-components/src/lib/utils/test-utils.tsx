/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
