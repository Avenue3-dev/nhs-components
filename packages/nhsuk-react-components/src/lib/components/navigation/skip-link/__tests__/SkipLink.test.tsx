import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import SkipLink from '..'

describe('SkipLink', () => {
  test('should match the snapshot', () => {
    render(<SkipLink data-testid="skip-link" />)
    expect(screen.getAllByTestId('skip-link')).toMatchSnapshot()
  })

  test('should render the correct href and content', () => {
    render(
      <SkipLink data-testid="skip-link" href="#maincontent">
        Skip to main content
      </SkipLink>,
    )
    expect(screen.getByTestId('skip-link')).toMatchSnapshot()
  })
})
