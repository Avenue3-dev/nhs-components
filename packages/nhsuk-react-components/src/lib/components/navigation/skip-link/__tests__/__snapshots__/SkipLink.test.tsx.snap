// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SkipLink > should match the snapshot 1`] = `
[
  <a
    class="nhsuk-skip-link"
    data-testid="skip-link"
    href="#maincontent"
  >
    Skip to main content
  </a>,
]
`;

exports[`SkipLink > should render the correct href and content 1`] = `
<a
  class="nhsuk-skip-link"
  data-testid="skip-link"
  href="#maincontent"
>
  Skip to main content
</a>
`;
